tools/cwtool/TxSimulator.hs

```diff
@@ -2,113 +2,242 @@
+                  txc noSPVSupport cmd (initGas cmdPwt) mc
+              T.putStrLn (encodeToText cr)
+        Nothing -> do -- blocks simulation
+          paydb <- newPayloadDb
+          withRocksDb "txsim-rocksdb" modernDefaultOptions $ \rdb ->
+            withBlockHeaderDb rdb ver cid $ \bdb -> do
+              let pse = PactServiceEnv Nothing cpe paydb bdb getGasModel readRewards 0 ferr
+                        ver True False logger gasLogger (pactLoggers cwLogger) False 1 defaultBlockGasLimit
+                  pss = PactServiceState Nothing mempty (ParentHeader parent) noSPVSupport
+              evalPactServiceM pss pse $ doBlock True parent (zip hdrs pwos)
+
+
 
 
   where
+
     cwLogger = genericLogger Debug T.putStrLn
     initGas cmd = initialGasOf (_cmdPayload cmd)
     logger = newLogger (pactLoggers cwLogger) "TxSimulator"
     gasLogger = Nothing
     txContext parent cmd = TxContext (ParentHeader parent) $ publicMetaOf cmd
+    ferr e _ = throwM e
+
+    doBlock
+        :: PayloadCasLookup cas
+        => Bool
+        -> BlockHeader
+        -> [(BlockHeader,PayloadWithOutputs)]
+        -> PactServiceM cas ()
+    doBlock _ _ [] = return ()
+    doBlock initMC parent ((hdr,pwo):rest) = do
+      !cp <- getCheckpointer
+      pde'@(PactDbEnv' pde) <-
+          liftIO $ _cpRestore cp $ Just (succ (_blockHeight parent), _blockHash parent)
+      when initMC $ do
+        mc <- liftIO $ readInitModules logger pde (TxContext (ParentHeader parent) def)
+        updateInitCache mc
+      psParentHeader .= ParentHeader parent
+      liftIO (spvSim sc hdr pwo) >>= assign psSpvSupport
+      _r <- trace (logFunction cwLogger) "execBlock" () 1 $
+          execBlock hdr (payloadWithOutputsToPayloadData pwo) pde'
+      liftIO $ _cpSave cp (_blockHash hdr)
+      doBlock False hdr rest
 
+-- | Block-scoped SPV mock by matching cont proofs to payload txs.
+-- Transactions are eliminated by searching for first matching proof in input;
+-- there should always be as many exact matches as proofs.
+-- TODO Heuristic for failed SPV needs to interrogate failures from outputs.
+spvSim :: SimConfig -> BlockHeader -> PayloadWithOutputs -> IO SPVSupport
+spvSim sc bh pwo = do
+  mv <- newMVar (V.toList (_payloadWithOutputsTransactions pwo))
+  return $ SPVSupport (_spvSupport noSPVSupport) (go mv)
+  where
+    go mv cp = modifyMVar mv $ searchOuts cp
+    searchOuts _ [] = return ([],Left "spv: proof not found")
+    searchOuts cp@(ContProof pf) ((Transaction ti,TransactionOutput o):txs) =
+      case codecDecode (chainwebPayloadCodec (Just (scVersion sc,_blockHeight bh))) ti of
+        Left {} -> internalError "input decode failed"
+        Right cmd -> case _pPayload $ payloadObj $ _cmdPayload cmd of
+          Continuation cm | _cmProof cm == Just cp -> do
+            cr :: CommandResult Hash <- decodeStrictOrThrow o
+            case _pactResult (_crResult cr) of
+              _ -> do -- TODO failed SPV error heuristic
```

Comment:
on `Left` here we could make it less opaque and error out iwth something like "tx failure for spv proofs unsupported at this time" rather than comment.

Reply:
I'll just remove it, we're covering every error seen so far with just the ones thrown here, and it will be faster not to decode outputs.